service/service_construction.go

```diff
@@ -545,6 +552,55 @@ func createUnwrapOps(tx transaction) ([]*types.Operation, *string, *types.Error)
 	return ops, &checkFrom, nil
 }
 
+func createGenericContractCallOps(tx transaction) ([]*types.Operation, *string, *types.Error) {
+	value := tx.Value
+
+	// Ensure valid from address
+	checkFrom, ok := ChecksumAddress(tx.From)
+	if !ok {
+		return nil, nil, WrapError(
+			ErrInvalidInput,
+			fmt.Errorf("%s is not a valid address", tx.From),
+		)
+	}
+
+	// Ensure valid to address
+	checkTo, ok := ChecksumAddress(tx.To)
+	if !ok {
+		return nil, nil, WrapError(ErrInvalidInput, fmt.Errorf("%s is not a valid address", tx.To))
+	}
+
+	ops := []*types.Operation{
```

Comment:
It seems kind of weird to me that the `data` for the call isn't in `operation.Metadata`?

https://www.rosetta-api.org/docs/models/Operation.html

Reply:
is this just because of the way you are indexing these operations?

Reply:
yeah, we don't put contract call data into the operation object, we put it in the `options` struct which is to follow the existing design like in this [rosetta geth sdk implementation](https://github.com/coinbase/rosetta-geth-sdk/blob/master/services/construction/preprocess.go#L155C11-L155C23)

Reply:
Got it. I think that gets away a bit from the intent of `Operations` (which should have full-fidelity of on-chain interaction) but that's not a convo for this forum.